**IST664 - NLP Lab Week 1**

Before getting started on this lab, make sure you have reviewed the introductory Python notebook (unless you are already experienced with Python). In this first section, we will use the Python “import” statement to load the NLTK package. This contains lots of text examples and unique capabilities for basic syntactic analysis of text. Whenever you see a code block containing only a comment (#) follow the instructions to add your own code.

Among the loaded objects, NLTK includes 9 of the text examples available from the corpora package (only a small number of them!). It has used the variable names text1 through text9 for these examples, and already assigned them values. The variables sent1 through sent9 have been set to be a list of tokens of the first sentence of each text.



In [2]:
# IST664 - NLP Lab Week 1
# This notebook has small examples that are meant to be run step by step.

import nltk
nltk.download('gutenberg')
nltk.download('genesis')
nltk.download('inaugural')
nltk.download('treebank')
nltk.download('nps_chat')
nltk.download('webtext')
from nltk.book import * # This may throw errors, looking for other downloads
# Just add them into the list of downloads above.



[nltk_data] Downloading package gutenberg to /root/nltk_data...
[nltk_data]   Unzipping corpora/gutenberg.zip.
[nltk_data] Downloading package genesis to /root/nltk_data...
[nltk_data]   Unzipping corpora/genesis.zip.
[nltk_data] Downloading package inaugural to /root/nltk_data...
[nltk_data]   Unzipping corpora/inaugural.zip.
[nltk_data] Downloading package treebank to /root/nltk_data...
[nltk_data]   Unzipping corpora/treebank.zip.
[nltk_data] Downloading package nps_chat to /root/nltk_data...
[nltk_data]   Unzipping corpora/nps_chat.zip.
[nltk_data] Downloading package webtext to /root/nltk_data...
[nltk_data]   Unzipping corpora/webtext.zip.


*** Introductory Examples for the NLTK Book ***
Loading text1, ..., text9 and sent1, ..., sent9
Type the name of the text or sentence to view it.
Type: 'texts()' or 'sents()' to list the materials.
text1: Moby Dick by Herman Melville 1851
text2: Sense and Sensibility by Jane Austen 1811
text3: The Book of Genesis
text4: Inaugural Address Corpus
text5: Chat Corpus
text6: Monty Python and the Holy Grail
text7: Wall Street Journal
text8: Personals Corpus
text9: The Man Who Was Thursday by G . K . Chesterton 1908


In [44]:
print(text2)

<Text: Sense and Sensibility by Jane Austen 1811>


In [62]:
# Now you try it by repeating this command, but using the imported
# data object text2. Write a comment indicating what text2 contains.


# 1.1: Your description of text2: Sense and Sensibility by Jane Austen 1811
print(text2)

<Text: Sense and Sensibility by Jane Austen 1811>


In [5]:
type(text1) # The type() command reveals the data type of an object

nltk.text.Text

In [6]:
dir(sent1) # These are the attributes of the object
# Check out how the end of the list shows all of the methods

['__add__',
 '__class__',
 '__class_getitem__',
 '__contains__',
 '__delattr__',
 '__delitem__',
 '__dir__',
 '__doc__',
 '__eq__',
 '__format__',
 '__ge__',
 '__getattribute__',
 '__getitem__',
 '__gt__',
 '__hash__',
 '__iadd__',
 '__imul__',
 '__init__',
 '__init_subclass__',
 '__iter__',
 '__le__',
 '__len__',
 '__lt__',
 '__mul__',
 '__ne__',
 '__new__',
 '__reduce__',
 '__reduce_ex__',
 '__repr__',
 '__reversed__',
 '__rmul__',
 '__setattr__',
 '__setitem__',
 '__sizeof__',
 '__str__',
 '__subclasshook__',
 'append',
 'clear',
 'copy',
 'count',
 'extend',
 'index',
 'insert',
 'pop',
 'remove',
 'reverse',
 'sort']

In [7]:
# Given the list above, we can find out more things about each attribute
getattr(text1, "concordance")

# <bound method Text.concordance of <Text: Moby Dick by Herman Melville 1851>>
# A bound method is a function that is an attribute of a class and can be
# called using an expression like : text1.concordance()
# See the next section for more information.

<bound method Text.concordance of <Text: Moby Dick by Herman Melville 1851>>

In [43]:
type(sent1) # What type is sent1
# Now you add a line of code that shows the attributes of sent1

# 1.2: Attributes of sent1

dir(sent1)


['__add__',
 '__class__',
 '__class_getitem__',
 '__contains__',
 '__delattr__',
 '__delitem__',
 '__dir__',
 '__doc__',
 '__eq__',
 '__format__',
 '__ge__',
 '__getattribute__',
 '__getitem__',
 '__gt__',
 '__hash__',
 '__iadd__',
 '__imul__',
 '__init__',
 '__init_subclass__',
 '__iter__',
 '__le__',
 '__len__',
 '__lt__',
 '__mul__',
 '__ne__',
 '__new__',
 '__reduce__',
 '__reduce_ex__',
 '__repr__',
 '__reversed__',
 '__rmul__',
 '__setattr__',
 '__setitem__',
 '__sizeof__',
 '__str__',
 '__subclasshook__',
 'append',
 'clear',
 'copy',
 'count',
 'extend',
 'index',
 'insert',
 'pop',
 'remove',
 'reverse',
 'sort']

In [9]:
print(sent1) # Lists are surrounded by [ and ]

['Call', 'me', 'Ishmael', '.']


In [10]:
sent1.sort() # This reorders the list alphabetically
print(sent1)

['.', 'Call', 'Ishmael', 'me']


In [11]:
sent1.remove('.') # We can delete items by name
print(sent1)

['Call', 'Ishmael', 'me']


In [12]:
sent1.append('.') # And we can add an item onto the end of the list
print(sent1)

['Call', 'Ishmael', 'me', '.']


In [13]:
# The join method can be applied to
sent1text = ' '.join(sent1) # You can glue elements of the list back together
print(sent1text)

Call Ishmael me .


In [14]:
# Now you glue sent2 back together with spaces in between and then
# print it (two lines of code).

# 1.3: Reglue sent2 tokens

sent2text = ' '.join(sent2) # You can glue elements of the list back together
print(sent2text)


The family of Dashwood had long been settled in Sussex .


**Searching**

The text data structure from nltk has a number of bound methods. One is called concordance(), and it will search for any word that you give to the function and print the occurrences and some surrounding context. This function is just a diagnostic convenience, as you cannot save the results for later use. A more flexible way of searching uses the findall() method. Another function is “similar” which finds all the words that are used in the same context as the one given, where the context is the word before and the word after. We can also examine the contexts that are shared by two or more words, such as “monstrous” and “very” by using common_contexts. We have to enclose these words by square brackets as well as parentheses, and separate them with a comma. The output of common_contexts is a pair of words that surrounds both of the target words provided. For example if the target is very and the function returns a_pretty, the fragment in question would be "a very pretty".

In [15]:
# searching text
text1.concordance('morning')


Displaying 25 of 74 matches:
eed her reported in the offing this morning ; a three years ' voyage , and a fu
or who could tell but what the next morning , so soon as I popped out of the ro
d Saturday night , or rather Sunday morning , in peddling his head around this 
 The Counterpane . Upon waking next morning about daylight , I found Queequeg '
ed away from me ; but waking in the morning , I shudderingly remembered it all 
t the first go off of a bitter cold morning . Seeing , now , that there were no
 wash himself . At that time in the morning any Christian would have washed his
et , all wearing monkey jackets for morning gowns . You could pretty plainly te
I did not . Returning from my first morning stroll , I again sallied out upon t
d thrown over me upon waking in the morning , I thought this indifference of hi
and chat over old times till nearly morning . Thus , then , in our hearts ' hon
ing . CHAPTER 13 Wheelbarrow . Next morning , Monday , after disposing of the e
d not at al

In [16]:
# A more flexible capability is provided by findall(), which can use a
# "regular expression" to find instances. Single tokens must be surrounded
# by angle brackets < >
text1.findall('<morning>')

morning; morning; morning; morning; morning; morning; morning;
morning; morning; morning; morning; morning; morning; morning;
morning; morning; morning; morning; morning; morning; morning;
morning; morning; morning; morning; morning; morning; morning;
morning; morning; morning; morning; morning; morning; morning;
morning; morning; morning; morning; morning; morning; morning;
morning; morning; morning; morning; morning; morning; morning;
morning; morning; morning; morning; morning; morning; morning;
morning; morning; morning; morning; morning; morning; morning;
morning; morning; morning; morning; morning


In [17]:
# Note that these token specifications are case sensitive
text1.findall('<Morning><to><ye>')

Morning to ye; Morning to ye; Morning to ye; Morning to ye; Morning to
ye


In [18]:
# We can also include two wildcard tokens to find out what comes next
text1.findall('<Morning><to><ye><.*><.*>')

Morning to ye , shipmates; Morning to ye , shipmates; Morning to ye ,
shipmates; Morning to ye ." Once; Morning to ye ! morning


In [19]:
text1.similar('morning') # These words are used in the same context as morning

sea time whale ship man water voyage same moment world while hand word
case lord day boat land days other


In [20]:
text2.similar('morning') # And here's the same thing for the second book

day house time room world place matter cottage evening point country
door and rest thing moment latter subject better letter


In [21]:
# The use of the square brackets makes a list from the two words
text2.common_contexts(["morning", "evening"])

the_was this_i the_of the_marianne the_for the_before


In [63]:
# Now you try it by repeating the common_contexts command, but using the
# data object text2. Use the terms morning and evening.

text2.common_contexts(["morning", "evening"])

# 1.4: Your explanation of the output: In text2 we will find sentences like "the morning of"/ "the evening of" / "the evening was" / "the morning for"

the_was this_i the_of the_marianne the_for the_before


**Word Counting**

A natural question to ask in the analysis of a text is: how many words are there in the text? While it is straightforward for human beings to answer the question, it is not so for the computers. As a starting point, we can use Python function len to first count the total number of words and punctuation symbols that appear.

In [23]:
# Counting the total number of tokens
len(text1)

260819

How to interpret this number 260819? This means that there are 260,819 words and punctuation symbols, or "tokens." A token is the technical name for a sequence of characters which is treated as a group by the program. Each text from the books was separated into a list of tokens, and this is one of the first NLP processing steps.

Now this is the total number of tokens, and we might also want to find out how many distinct words there are, not counting repetitions. The Python “set” function removes the repetitions, and we can apply the “sorted” function to that, returning the resulted sorted list of tokens.

In [24]:
sorted(set(text1))[:10] # Review the first 10 of an alphabetized set of tokens

['!', '!"', '!"--', "!'", '!\'"', '!)', '!)"', '!*', '!--', '!--"']

In [25]:
# Let's look at a few more, after we've gotten through the punctuation, numbers
# and other stuff that appears early in the sort.
sorted(set(text1))[400:410]

['Alone',
 'Alps',
 'Already',
 'Also',
 'Am',
 'Ambergriese',
 'Ambergris',
 'Amelia',
 'America',
 'American']

In [26]:
# The set() function creates an unordered collection
# of elements where each element is unique.
len(sorted(set(text1))) # Total number of tokens

19317

In [27]:
len(set(text1)) / len(text1) # The size of the token set versus total tokens

0.07406285585022564

In [28]:
text1.count("the") # Count occurrences of one word

13721

In [29]:
# Among tokens that are actual words, "the" is often one of the most common in
# modern English.
100 * text1.count('the')/len(text1) # As a percent of all tokens

5.260736372733581

In [30]:
100 * text1.count('and')/len(text1) # A more common word

2.3096476867099405

In [31]:
100 * text1.count('morning')/len(text1) # Our example from above

0.02607172023510557

In [32]:
# This uses a list comprehension to cycle through all of the words in text1
sixlettertokens = [w for w in text1 if len(w) >=6]
len(sixlettertokens) # Count how many tokens are 6 letters or more

55208

In [33]:
len(text2) # Now repeat the analyses above for text2

141576

In [34]:
# For text2, calculate the following quantities:

# 1.5: The size of the token set versus total tokens (expressed as a proportion)
print("The size of the token set versus total tokens = ",len(set(text2)) / len(text2))


# 1.6: The occurrences of the word 'the' as a percentage of all tokens
print("The occurrences of the word 'the' as a percentage of all tokens = ",100 * text2.count('the')/len(text2))

# 1.7: The occurrences of the word 'morning' as a percentage of all tokens
print("The occurrences of the word 'morning' as a percentage of all tokens = ",100 * text2.count('morning')/len(text2))


# 1.8: How many tokens are 7 letters or more
sevenlettertokens = [w for w in text1 if len(w) >6]
print("How many tokens are 7 letters or more = ",len(sevenlettertokens))

The size of the token set versus total tokens =  0.04826383002768831
The occurrences of the word 'the' as a percentage of all tokens =  2.727157145278861
The occurrences of the word 'morning' as a percentage of all tokens =  0.061451093405661975
How many tokens are 7 letters or more =  38097


In [35]:
# Here we are creating a custom function to integrate what we learned.
def day_word_analysis(input_text):
  """
  Find and print context for common time of day words.
  Print an analysis of the occurrence of these words.
  """
  input_text.concordance('morning', lines=3)
  input_text.concordance('afternoon', lines=3)

  m_count = input_text.count('morning')
  a_count = input_text.count('afternoon')

  print('Percentage of time of day words:',
        100*(m_count + a_count)/len(input_text))

In [36]:
day_word_analysis(text1)

Displaying 3 of 74 matches:
eed her reported in the offing this morning ; a three years ' voyage , and a fu
or who could tell but what the next morning , so soon as I popped out of the ro
d Saturday night , or rather Sunday morning , in peddling his head around this 
Displaying 3 of 10 matches:
-- GOLDSMITH TO JOHNSON . " In the afternoon we saw what was supposed to be a r
ulate the city of a dreamy Sabbath afternoon . Go from Corlears Hook to Coentie
h it was only two o ' clock in the afternoon of the 21st June , the longest day
Percentage of time of day words: 0.029905796740268154


In [42]:
# Now add 'evening' to day_word_analysis() and retest it.
# Make sure to test with text1 and text2.

def day_word_analysis(input_text):
  """
  Find and print context for common time of day words.
  Print an analysis of the occurrence of these words.
  """
  input_text.concordance('morning', lines=3)
  input_text.concordance('afternoon', lines=3)
  input_text.concordance('evening', lines=3)

  m_count = input_text.count('morning')
  a_count = input_text.count('afternoon')
  e_count = input_text.count('afternoon')


  print('Percentage of time of day words:',
        100*(m_count + a_count + e_count)/len(input_text))

#print("Text 1 results", day_word_analysis(text1))
#print("Text 2 results", day_word_analysis(text2))

# 1.9: What do the results show?

day_word_analysis(text1)
print("\n")
day_word_analysis(text2)

Displaying 3 of 74 matches:
eed her reported in the offing this morning ; a three years ' voyage , and a fu
or who could tell but what the next morning , so soon as I popped out of the ro
d Saturday night , or rather Sunday morning , in peddling his head around this 
Displaying 3 of 10 matches:
-- GOLDSMITH TO JOHNSON . " In the afternoon we saw what was supposed to be a r
ulate the city of a dreamy Sabbath afternoon . Go from Corlears Hook to Coentie
h it was only two o ' clock in the afternoon of the 21st June , the longest day
Displaying 3 of 18 matches:
I resolved to spend the rest of the evening as a looker on . Presently a riotin
en only glows to be looked at ; the evening shades and phantoms gathering round
 them stay . He then went about his evening prayers , took out his idol , and r
Percentage of time of day words: 0.03373987324543074


Displaying 3 of 87 matches:
e had been to several families that morning in hopes of procuring some addition
 together in the course of the la

In [60]:

# 1.10: Create a new function that discovers a different class of words
# such as the names of characters in the story. Calculate the frequency of
# occurrence of all the names together and also report which is the most
# frequently used name.


def names_analysis(input_text):
  """
  Find and print context for common names used in the text.
  Print an analysis of the occurrence of these words.
  """
  input_text.concordance('Marianne', lines=3)
  input_text.concordance('Margaret', lines=3)
  input_text.concordance('Elinor', lines=3)

  marianne_count = input_text.count('Marianne')
  margaret_count = input_text.count('Margaret')
  elinor_count = input_text.count('Elinor')


  print('\nPercentage of time of Marianne:',
        100*(marianne_count)/len(input_text))
  print('\nPercentage of time of Margaret:',
        100*(margaret_count)/len(input_text))
  print('\nPercentage of time of Elinor:',
        100*(elinor_count)/len(input_text))

names_analysis(text2)



Displaying 3 of 566 matches:
 had resolved never to be taught . Marianne ' s abilities were , in many respe
had already imbibed a good deal of Marianne ' s romance , without having much 
hing . " In a few months , my dear Marianne ." said she , " Elinor will , in a
Displaying 3 of 36 matches:
urage her to similar forbearance . Margaret , the other sister , was a good - 
ing in the opinion of Marianne and Margaret an absolute old bachelor , for he 
ne of these hills did Marianne and Margaret one memorable morning direct their
Displaying 3 of 685 matches:
avoid a breach with their brother . Elinor , this eldest daughter , whose advi
, in many respects , quite equal to Elinor ' s . She was sensible and clever ;
d her mother was strikingly great . Elinor saw , with concern , the excess of 

Percentage of time of Marianne: 0.39978527434028366

Percentage of time of Margaret: 0.02542803865061875

Percentage of time of Elinor: 0.48313273436175624
